File: apps/ai-service/src/utils/logger.py
import logging
import json
import sys
import os
from datetime import datetime
from typing import Dict, Any, Optional
from pathlib import Path

# Create logs directory if it doesn't exist
logs_dir = Path("logs")
logs_dir.mkdir(exist_ok=True)

class StructuredFormatter(logging.Formatter):
    """Custom formatter for structured JSON logging"""
    
    def __init__(self, service_name: str = "lusilearn-ai"):
        self.service_name = service_name
        self.environment = os.getenv("ENVIRONMENT", "development")
        super().__init__()
    
    def format(self, record: logging.LogRecord) -> str:
        log_entry = {
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "level": record.levelname,
            "service": self.service_name,
            "environment": self.environment,
            "message": record.getMessage(),
            "module": record.module,
            "function": record.funcName,
            "line": record.lineno
        }
        
        # Add extra fields if present
        if hasattr(record, 'user_id'):
            log_entry['userId'] = record.user_id
        if hasattr(record, 'request_id'):
            log_entry['requestId'] = record.request_id
        if hasattr(record, 'duration'):
            log_entry['duration'] = record.duration
        if hasattr(record, 'model'):
            log_entry['model'] = record.model
        if hasattr(record, 'tokens'):
            log_entry['tokens'] = record.tokens
        if hasattr(record, 'cost'):
            log_entry['cost'] = record.cost
        
        # Add exception info if present
        if record.exc_info:
            log_entry['exception'] = self.formatException(record.exc_info)
        
        # Add any additional fields from extra parameter
        if hasattr(record, 'extra_fields'):
            log_entry.update(record.extra_fields)
        
        return json.dumps(log_entry)

class AIServiceLogger:
    """Centralized logging service for AI operations"""
    
    def __init__(self):
        self.logger = logging.getLogger("ai_service")
        self.performance_logger = logging.getLogger("ai_performance")
        self.cost_logger = logging.getLogger("ai_cost")
        self.security_logger = logging.getLogger("ai_security")
        
        self._setup_loggers()
    
    def _setup_loggers(self):
        """Setup all loggers with appropriate handlers and formatters"""
        log_level = os.getenv("LOG_LEVEL", "INFO").upper()
        
        # Main application logger
        self._setup_logger(
            self.logger,
            log_level,
            "logs/ai-service.log",
            "logs/ai-error.log"
        )
        
        # Performance logger
        self._setup_logger(
            self.performance_logger,
            "INFO",
            "logs/ai-performance.log"
        )
        
        # Cost tracking logger
        self._setup_logger(
            self.cost_logger,
            "INFO",
            "logs/ai-cost.log"
        )
        
        # Security logger
        self._setup_logger(
            self.security_logger,
            "INFO",
            "logs/ai-security.log"
        )
    
    def _setup_logger(self, logger: logging.Logger, level: str, 
                     info_file: str, error_file: Optional[str] = None):
        """Setup individual logger with handlers"""
        logger.setLevel(getattr(logging, level))
        
        # Remove existing handlers
        for handler in logger.handlers[:]:
            logger.removeHandler(handler)
        
        formatter = StructuredFormatter()
        
        # File handler for all logs
        file_handler = logging.FileHandler(info_file)
        file_handler.setLevel(logging.INFO)
        file_handler.setFormatter(formatter)
        logger.addHandler(file_handler)
        
        # Separate error file handler if specified
        if error_file:
            error_handler = logging.FileHandler(error_file)
            error_handler.setLevel(logging.ERROR)
            error_handler.setFormatter(formatter)
            logger.addHandler(error_handler)
        
        # Console handler for development
        if os.getenv("ENVIRONMENT") != "production":
            console_handler = logging.StreamHandler(sys.stdout)
            console_handler.setLevel(logging.DEBUG)
            
            # Simple formatter for console
            console_formatter = logging.Formatter(
                '%(asctime)s [%(name)s] %(levelname)s: %(message)s'
            )
            console_handler.setFormatter(console_formatter)
            logger.addHandler(console_handler)
        
        # Prevent propagation to root logger
        logger.propagate = False
    
    def info(self, message: str, **kwargs):
        """Log info message with optional context"""
        self.logger.info(message, extra={'extra_fields': kwargs})
    
    def error(self, message: str, **kwargs):
        """Log error message with optional context"""
        self.logger.error(message, extra={'extra_fields': kwargs})
    
    def warning(self, message: str, **kwargs):
        """Log warning message with optional context"""
        self.logger.warning(message, extra={'extra_fields': kwargs})
    
    def debug(self, message: str, **kwargs):
        """Log debug message with optional context"""
        self.logger.debug(message, extra={'extra_fields': kwargs})
    
    def log_performance(self, operation: str, duration: float, 
                       user_id: Optional[str] = None, **kwargs):
        """Log performance metrics"""
        self.performance_logger.info(
            f"AI Operation: {operation}",
            extra={
                'extra_fields': {
                    'operation': operation,
                    'duration': duration,
                    'user_id': user_id,
                    **kwargs
                }
            }
        )
    
    def log_ai_usage(self, model: str, operation: str, tokens: int, 
                    cost: float, user_id: Optional[str] = None, **kwargs):
        """Log AI API usage and costs"""
        self.cost_logger.info(
            f"AI Usage: {operation} with {model}",
            extra={
                'extra_fields': {
                    'model': model,
                    'operation': operation,
                    'tokens': tokens,
                    'cost': cost,
                    'user_id': user_id,
                    **kwargs
                }
            }
        )
    
    def log_security_event(self, event: str, user_id: Optional[str] = None, 
                          ip: Optional[str] = None, **kwargs):
        """Log security-related events"""
        self.security_logger.info(
            f"Security Event: {event}",
            extra={
                'extra_fields': {
                    'event': event,
                    'user_id': user_id,
                    'ip': ip,
                    **kwargs
                }
            }
        )
    
    def log_model_inference(self, model: str, input_tokens: int, 
                           output_tokens: int, latency: float,
                           user_id: Optional[str] = None, **kwargs):
        """Log model inference metrics"""
        total_tokens = input_tokens + output_tokens
        
        self.performance_logger.info(
            f"Model Inference: {model}",
            extra={
                'extra_fields': {
                    'model': model,
                    'input_tokens': input_tokens,
                    'output_tokens': output_tokens,
                    'total_tokens': total_tokens,
                    'latency': latency,
                    'tokens_per_second': total_tokens / latency if latency > 0 else 0,
                    'user_id': user_id,
                    **kwargs
                }
            }
        )
    
    def log_recommendation_request(self, user_id: str, content_type: str,
                                 num_recommendations: int, duration: float, **kwargs):
        """Log content recommendation requests"""
        self.performance_logger.info(
            f"Content Recommendation Request",
            extra={
                'extra_fields': {
                    'user_id': user_id,
                    'content_type': content_type,
                    'num_recommendations': num_recommendations,
                    'duration': duration,
                    **kwargs
                }
            }
        )
    
    def log_learning_path_generation(self, user_id: str, subject: str,
                                   num_objectives: int, duration: float, **kwargs):
        """Log learning path generation"""
        self.performance_logger.info(
            f"Learning Path Generation",
            extra={
                'extra_fields': {
                    'user_id': user_id,
                    'subject': subject,
                    'num_objectives': num_objectives,
                    'duration': duration,
                    **kwargs
                }
            }
        )

# Global logger instance
ai_logger = AIServiceLogger()

# Convenience functions for backward compatibility
def log_info(message: str, **kwargs):
    ai_logger.info(message, **kwargs)

def log_error(message: str, **kwargs):
    ai_logger.error(message, **kwargs)

def log_warning(message: str, **kwargs):
    ai_logger.warning(message, **kwargs)

def log_debug(message: str, **kwargs):
    ai_logger.debug(message, **kwargs)

def log_performance(operation: str, duration: float, **kwargs):
    ai_logger.log_performance(operation, duration, **kwargs)

def log_ai_usage(model: str, operation: str, tokens: int, cost: float, **kwargs):
    ai_logger.log_ai_usage(model, operation, tokens, cost, **kwargs)

def log_security_event(event: str, **kwargs):
    ai_logger.log_security_event(event, **kwargs)

File: apps/api/src/utils/logger.ts
import winston from 'winston';
import path from 'path';
import fs from 'fs';

const logLevel = process.env.LOG_LEVEL || 'info';
const serviceName = process.env.SERVICE_NAME || 'lusilearn-api';
const environment = process.env.NODE_ENV || 'development';

// Ensure logs directory exists
const logsDir = path.join(process.cwd(), 'logs');
if (!fs.existsSync(logsDir)) {
  fs.mkdirSync(logsDir, { recursive: true });
}

// Custom format for structured logging
const structuredFormat = winston.format.combine(
  winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss.SSS' }),
  winston.format.errors({ stack: true }),
  winston.format.printf((info) => {
    const { timestamp, level, message, service, userId, requestId, duration, ...meta } = info;
    const logEntry = {
      timestamp,
      level,
      service,
      environment,
      message,
      ...(userId ? { userId } : {}),
      ...(requestId ? { requestId } : {}),
      ...(duration ? { duration } : {}),
      ...meta
    };
    return JSON.stringify(logEntry);
  })
);

// Performance logging format
const performanceFormat = winston.format.combine(
  winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss.SSS' }),
  winston.format.printf((info) => {
    const { timestamp, level, message, service, method, url, statusCode, duration, userId, ...meta } = info;
    const logEntry = {
      timestamp,
      level,
      service,
      environment,
      type: 'performance',
      method,
      url,
      statusCode,
      duration,
      ...(userId ? { userId } : {}),
      message,
      ...meta
    };
    return JSON.stringify(logEntry);
  })
);

// Security logging format
const securityFormat = winston.format.combine(
  winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss.SSS' }),
  winston.format.printf((info) => {
    const { timestamp, level, message, service, userId, ip, userAgent, action, ...meta } = info;
    const logEntry = {
      timestamp,
      level,
      service,
      environment,
      type: 'security',
      userId,
      ip,
      userAgent,
      action,
      message,
      ...meta
    };
    return JSON.stringify(logEntry);
  })
);

// Main application logger
export const logger = winston.createLogger({
  level: logLevel,
  format: structuredFormat,
  defaultMeta: { service: serviceName },
  transports: [
    new winston.transports.File({ 
      filename: path.join(logsDir, 'error.log'), 
      level: 'error',
      maxsize: 50 * 1024 * 1024, // 50MB
      maxFiles: 5,
      tailable: true
    }),
    new winston.transports.File({ 
      filename: path.join(logsDir, 'combined.log'),
      maxsize: 100 * 1024 * 1024, // 100MB
      maxFiles: 10,
      tailable: true
    }),
  ],
});

// Performance logger for API metrics
export const performanceLogger = winston.createLogger({
  level: 'info',
  format: performanceFormat,
  defaultMeta: { service: serviceName },
  transports: [
    new winston.transports.File({ 
      filename: path.join(logsDir, 'performance.log'),
      maxsize: 100 * 1024 * 1024, // 100MB
      maxFiles: 5,
      tailable: true
    }),
  ],
});

// Security logger for authentication and authorization events
export const securityLogger = winston.createLogger({
  level: 'info',
  format: securityFormat,
  defaultMeta: { service: serviceName },
  transports: [
    new winston.transports.File({ 
      filename: path.join(logsDir, 'security.log'),
      maxsize: 50 * 1024 * 1024, // 50MB
      maxFiles: 10,
      tailable: true
    }),
  ],
});

// AI cost and usage logger
export const aiLogger = winston.createLogger({
  level: 'info',
  format: winston.format.combine(
    winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss.SSS' }),
    winston.format.printf((info) => {
      const { timestamp, level, message, service, userId, model, tokens, cost, operation, ...meta } = info;
      const logEntry = {
        timestamp,
        level,
        service,
        environment,
        type: 'ai_usage',
        userId,
        model,
        tokens,
        cost,
        operation,
        message,
        ...meta
      };
      return JSON.stringify(logEntry);
    })
  ),
  defaultMeta: { service: serviceName },
  transports: [
    new winston.transports.File({ 
      filename: path.join(logsDir, 'ai-usage.log'),
      maxsize: 50 * 1024 * 1024, // 50MB
      maxFiles: 5,
      tailable: true
    }),
  ],
});

// User analytics logger
export const analyticsLogger = winston.createLogger({
  level: 'info',
  format: winston.format.combine(
    winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss.SSS' }),
    winston.format.printf((info) => {
      const { timestamp, level, message, service, userId, sessionId, event, properties, ...meta } = info;
      const logEntry = {
        timestamp,
        level,
        service,
        environment,
        type: 'analytics',
        userId,
        sessionId,
        event,
        properties,
        message,
        ...meta
      };
      return JSON.stringify(logEntry);
    })
  ),
  defaultMeta: { service: serviceName },
  transports: [
    new winston.transports.File({ 
      filename: path.join(logsDir, 'analytics.log'),
      maxsize: 100 * 1024 * 1024, // 100MB
      maxFiles: 10,
      tailable: true
    }),
  ],
});

// Console logging for development
if (environment !== 'production') {
  const consoleTransport = new winston.transports.Console({
    format: winston.format.combine(
      winston.format.colorize(),
      winston.format.timestamp({ format: 'HH:mm:ss' }),
      winston.format.printf((info) => {
        const { timestamp, level, message, service, userId, requestId, ...meta } = info;
        const metaStr = Object.keys(meta).length ? ` ${JSON.stringify(meta)}` : '';
        const userStr = userId ? ` [User: ${userId}]` : '';
        const reqStr = requestId ? ` [Req: ${(requestId as string).slice(0, 8)}]` : '';
        return `${timestamp} [${service}] ${level}:${userStr}${reqStr} ${message}${metaStr}`;
      })
    )
  });

  logger.add(consoleTransport);
  performanceLogger.add(consoleTransport);
  securityLogger.add(consoleTransport);
  aiLogger.add(consoleTransport);
  analyticsLogger.add(consoleTransport);
}

// Helper functions for structured logging
export const logWithContext = (level: string, message: string, context: Record<string, any> = {}) => {
  logger.log(level, message, context);
};

export const logPerformance = (method: string, url: string, statusCode: number, duration: number, userId?: string, additionalData?: Record<string, any>) => {
  performanceLogger.info('API Request', {
    method,
    url,
    statusCode,
    duration,
    userId,
    ...additionalData
  });
};

export const logSecurity = (action: string, userId: string, ip: string, userAgent: string, success: boolean, additionalData?: Record<string, any>) => {
  const level = success ? 'info' : 'warn';
  securityLogger.log(level, `Security event: ${action}`, {
    action,
    userId,
    ip,
    userAgent,
    success,
    ...additionalData
  });
};

export const logAIUsage = (operation: string, model: string, tokens: number, cost: number, userId?: string, additionalData?: Record<string, any>) => {
  aiLogger.info('AI API Usage', {
    operation,
    model,
    tokens,
    cost,
    userId,
    ...additionalData
  });
};

export const logAnalytics = (event: string, userId: string, sessionId: string, properties: Record<string, any> = {}) => {
  analyticsLogger.info('User Analytics Event', {
    event,
    userId,
    sessionId,
    properties
  });
};

export default logger;